### PR TITLE
Bkg callback

### PR DESCRIPTION
**Pull request by JamesKunstle**
Review note: 

There are three major changes in this PR-

1. All callbacks that create a figure become background_callbacks. The dcc.Graph component that they return to is wrapped in a dcc.Loading object. The callbacks themselves wait for the data they need (all of the data for the repo_choices input vector to be satisfied) to be populated in Redis.
2. Queries to Augur are Celery tasks that download all of the requested data at once and then chunk it post-download, saving it to Redis per-repo.
3. job_manager module has been dropped and the cache_manager module has been added.
4. Lots of miscellaneous formatting changes have been made because the 'black' for matter started working in VSCode.
5. switched from pipenv to requirements.txt because pipenv was giving a lot of trouble.
6. docker-compose added to replace /scripts/launch_app.sh

### CODE REVIEW

**Comment by JamesKunstle** [comment]
pre-commit failing because it doesn't like the 8knot module name. it thinks that it should follow pep8 and be all lower-case and non-numeric/non-special-characters

File: pages/overview.py [hunk 1]
```diff
@@ -3,10 +3,14 @@
...
 
 # import the visualization cards
-from .visualizations.overview.total_contributor_growth import gc_total_contributor_growth
+from .visualizations.overview.total_contributor_growth import (
+    gc_total_contributor_growth,
+)
 from .visualizations.overview.commits_over_time import gc_commits_over_time
 from .visualizations.overview.issues_over_time import gc_issues_over_time
-from .visualizations.overview.active_drifting_contributors import gc_active_drifting_contributors
+from .visualizations.overview.active_drifting_contributors import (
```
Comment (cdolfi) [inline pages/overview.py:11]: why this change?

File: pages/visualizations/chaoss/contributors_over_time.py [hunk 2]
```diff
@@ -171,7 +170,9 @@ def create_graph(repolist, timer_pings, contribs, interval):
 
     # df for drive by contributros in time interval
     df_drive = (
-        df_drive_temp.groupby(by=df_drive_temp.created.dt.to_period(interval))["cntrb_id"]
+        df_drive_temp.groupby(by=df_drive_temp.created.dt.to_period(interval))[
```
Comment (cdolfi) [inline pages/visualizations/chaoss/contributors_over_time.py:173]: I find the first option more readable, especially if you have pandas experience.

File: pages/visualizations/chaoss/contributors_over_time.py [hunk 3]
```diff
@@ -180,7 +181,9 @@ def create_graph(repolist, timer_pings, contribs, interval):
 
     # df for repeat contributors in time interval
     df_repeat = (
-        df_repeat_temp.groupby(by=df_repeat_temp.created.dt.to_period(interval))["cntrb_id"]
+        df_repeat_temp.groupby(by=df_repeat_temp.created.dt.to_period(interval))[
```
Comment (cdolfi) [inline pages/visualizations/chaoss/contributors_over_time.py:184]: same here

File: pages/visualizations/overview/active_drifting_contributors.py [hunk 4]
```diff
@@ -158,34 +152,34 @@ def toggle_popover_4(n, is_open):
...
+    logging.debug(f"ACTIVE_DRIFTING_CONTRIBUTOR_GROWTH_VIZ - START")
     start = time.perf_counter()
 
-    # create dataframe from record data
-    df = pd.DataFrame(results)
+    if df.empty:
+        return nodata_graph, False
 
     # test if there is data
     if df.empty:
```
Comment (cdolfi) [inline pages/visualizations/overview/active_drifting_contributors.py:185]: if df.empty statements in there twice

File: pages/visualizations/overview/issue_staleness.py [hunk 5]
```diff
@@ -155,38 +150,39 @@ def toggle_popover_issues(n, is_open):
...
     start = time.perf_counter()
+    logging.debug("ISSUES STALENESS - START")
 
-    # create dataframe from record data
-    df = pd.DataFrame(results)
+    if df.empty:
+        return nodata_graph, False
 
     # test if there is data
     if df.empty:
```
Comment (cdolfi) [inline pages/visualizations/overview/issue_staleness.py:183]: df.empty statements in there twice

File: pages/visualizations/overview/issues_over_time.py [hunk 6]
```diff
@@ -106,33 +98,74 @@ def toggle_popover_3(n, is_open):
     return is_open
 
 
+columns = ["1", "2", "3"]
```
Comment (cdolfi) [inline pages/visualizations/overview/issues_over_time.py:101]: what is this? seems like it shouldnt be here

File: pages/visualizations/overview/issues_over_time.py [hunk 7]
```diff
@@ -106,33 +98,74 @@ def toggle_popover_3(n, is_open):
...
+        "yanchor": "top",
+    },
+    font=dict(size=18, color="orange"),
+)
+
 # callback for issues over time graph
 @callback(
     Output("issues-over-time", "figure"),
-    Output("issues-over-time-timer", "n_intervals"),
+    # Output("issues-over-time-timer", "n_intervals"),
```
Comment (cdolfi) [inline pages/visualizations/overview/issues_over_time.py:144]: take out the comments if archived code

File: pages/visualizations/overview/pr_over_time.py [hunk 8]
```diff
@@ -109,30 +102,43 @@ def toggle_popover_7(n, is_open):
...
     start = time.perf_counter()
+    logging.debug("PULL REQUEST STALENESS - START")
+
+    # get all results from cache
+    results = cache.getm(func=prq, repos=repolist)
 
-    # create dataframe from record data
-    df = pd.DataFrame(results)
+    # deserialize results, create list of dfs
+    dfs = []
```
Comment (cdolfi) [inline pages/visualizations/overview/pr_over_time.py:127]: why is this different than the others?

File: queries/commits_query.py [hunk 9]
```diff
@@ -31,15 +44,32 @@ def commits_query(dbmc, repo):
                     FROM
                         repo r
                     JOIN commits c
-                    ON r.repo_id = c.repo_id
+                        ON r.repo_id = c.repo_id
                     WHERE
-                        c.repo_id = {repo}
+                        c.repo_id in ({str(repos)[1:-1]})
```
Comment (cdolfi) [inline queries/commits_query.py:49]: switching back to be by repo set not singular repo for caching?

File: queries/prs_query.py [hunk 10]
```diff
@@ -51,5 +64,22 @@ def prs_query(dbmc, repo):
...
+        c_df = pd.DataFrame(df_pr.loc[df_pr["id"] == r]).to_csv()
+
+        # add pickled dataframe to list of pickled objects
+        pic.append(c_df)
+
+    del df_pr
+
+    # store results in Redis
+    cm_o = cm()
+    ack = cm_o.setm(func=prs_query, repos=repos, datas=pic)
```
Comment (cdolfi) [inline queries/prs_query.py:82]: whats ack? comment on all queries

**Review by cdolfi — CHANGES_REQUESTED** [review]

File: pages/overview.py (hunk 1 above)
Comment (JamesKunstle) [reply pages/overview.py:11]: black formatter didn't like the line length I suppose. I didn't intentionally change that.

File: pages/visualizations/chaoss/contributors_over_time.py (hunk 2 above)
Comment (JamesKunstle) [reply pages/visualizations/chaoss/contributors_over_time.py:173]: I agree, like the other change, the 'black' formatter did it and will continue to do it.

File: pages/visualizations/chaoss/contributors_over_time.py (hunk 3 above)
Comment (JamesKunstle) [reply pages/visualizations/chaoss/contributors_over_time.py:184]: same

File: pages/visualizations/overview/active_drifting_contributors.py (hunk 4 above)
Comment (JamesKunstle) [reply pages/visualizations/overview/active_drifting_contributors.py:185]: ah good catch, copy-paste error on my part

File: pages/visualizations/overview/issue_staleness.py (hunk 5 above)
Comment (JamesKunstle) [reply pages/visualizations/overview/issue_staleness.py:183]: danke, copy-paste error.

File: pages/visualizations/overview/issues_over_time.py (hunk 6 above)
Comment (JamesKunstle) [reply pages/visualizations/overview/issues_over_time.py:101]: yeah that's an old artifact, I'll change that.

File: pages/visualizations/overview/issues_over_time.py (hunk 7 above)
Comment (JamesKunstle) [reply pages/visualizations/overview/issues_over_time.py:144]: ack

File: pages/visualizations/overview/pr_over_time.py (hunk 8 above)
Comment (JamesKunstle) [reply pages/visualizations/overview/pr_over_time.py:127]: missed it on refactor, thought I got all of them

File: queries/prs_query.py (hunk 10 above)
Comment (JamesKunstle) [reply queries/prs_query.py:82]: gotcha, 'ack' is a programming shorthand for 'acknowledgement' or 'acknowledged,' originating in embedded systems and networking. I used it reflexively but I'll change it.

File: pages/visualizations/chaoss/contributors_over_time.py (hunk 2 above)
Comment (cdolfi) [reply pages/visualizations/chaoss/contributors_over_time.py:173]: let the war begin lol

File: app.py [hunk 11]
```diff
@@ -133,6 +134,7 @@ def _project_list_query():
     [
         # componets to store data from queries
         dcc.Store(id="repo-choices", storage_type="session", data=[]),
+        dcc.Store(id="dead-end", storage_type="session", data=[]),
```
[the remark this answers is not in the document]
Comment (JamesKunstle) [reply app.py:137]: removed- from a past design that I forgot about

File: app_callbacks.py [hunk 12]
```diff
@@ -115,10 +122,65 @@ def update_output(n_clicks, value):
         return dash.exceptions.PreventUpdate, dash.exceptions.PreventUpdate
 
 
-@callback(Output("help-alert", "is_open"), Input("search-help", "n_clicks"), State("help-alert", "is_open"))
+@callback(
```
[the remark this answers is not in the document]
Comment (JamesKunstle) [reply app_callbacks.py:125]: yep

File: pages/visualizations/chaoss/contributors_over_time.py (hunk 2 above)
Comment (JamesKunstle) [reply pages/visualizations/chaoss/contributors_over_time.py:173]: i think I can **technically** change the line-length requirement for 'black' but tbh I'd rather keep it as standard as possible. I'm a pep8 / let-the-formatter-do-what-it-wants person

**Comment by JamesKunstle** [comment]
@cdolfi changes made to address your comments- let me know if there's anything else to address or please feel free to merge when convenient!

File: pages/overview.py [hunk 13]
```diff
@@ -3,10 +3,16 @@
 import dash_bootstrap_components as dbc
 
 # import the visualization cards
-from .visualizations.overview.total_contributor_growth import gc_total_contributor_growth
+from .visualizations.overview.total_contributor_growth import (
```
Comment (cdolfi) [inline pages/overview.py:6]: Lets add this one between lines 13-15

**Comment by JamesKunstle** [comment]
@cdolfi okay done!

**Review by cdolfi — APPROVED** [review]
